serverless-heaven/serverless-webpack
# Add Serverless Container Support

## What did you implement:

Closes https://github.com/serverless-heaven/serverless-webpack/issues/698



## How did you implement it:



I added another function to `validate.js` to get the actual handler name from the function definition, now that `serverless` rolled out `image` for custom runtimes. 

Review the following documentation: serverless documentation: https://www.serverless.com/blog/container-support-for-lambda. The only gap between feature support defined in the above article and this commit, is the ability to inherit the docker image's default `command` property. For example, `serverless-webpack` can't read the `CMD` property for the following Dockerfile:
```bash
# Dockerfile
FROM public.ecr.aws/lambda/nodejs:12
CMD "index.handler"
```
However you can simply set the `function.image.command[0]` property in `serverless.yml` as an override. I assume this would be the desired approach for most use cases, when sharing a docker image for all functions within the service. 
```YAML
# serverless.yml
service:
  ...
provider:
  ...
functions:
  myFunction1:
    image:
      name: public.ecr.aws/lambda/nodejs:12
      command:
        - 'index.handler1'
  myFunction2:
    image:
      name: public.ecr.aws/lambda/nodejs:12
      command:
        - 'index.handler2'
```

## How can we verify it:

* I added to the existing unit tests to check for a function definition with no `handler` property, and has `image.command` instead. Run `npm run test` to quickly validate test coverage:

![Screen Shot 2021-03-12 at 10 31 13 AM](https://user-images.githubusercontent.com/14983357/110961702-146edf00-831e-11eb-9014-9b95381692d3.png)

* On my branch I went into `serverless-webpack` and did a `yalc publish` to locally mock the `npm` package.
* I created a new serverless docker project with 

<img width="654" alt="Screen Shot 2021-03-12 at 10 47 11 AM" src="https://user-images.githubusercontent.com/14983357/110963630-4f721200-8320-11eb-8bc7-f906b7354a13.png">

* I opened the new project in my editor; I then ran the following:

```bash
npm init
npm install
npm install webpack serverless-webpack bestzip --dev
yalc link serverless-webpack # do a local install of my serverless-webpack branch
``` 

* I added a `plugins` section with `serverless-webpack`:

![Screen Shot 2021-03-12 at 10 52 08 AM](https://user-images.githubusercontent.com/14983357/110964361-1d14e480-8321-11eb-98e5-afc93473fe05.png)


* I added a `command` array to the existing `image` property for the function called `hello`:

![Screen Shot 2021-03-12 at 10 53 42 AM](https://user-images.githubusercontent.com/14983357/110964468-387fef80-8321-11eb-823a-e0911ec1cec9.png)

Note `app.js` contains a function called `handler`.
* I added the following to a new file called `webpack.config.js`
```bash
const path = require('path');
const slsw = require('serverless-webpack');

module.exports = {
  entry: slsw.lib.entries,
  target: 'node',
  optimization: {
    minimize: false,
  },
  performance: {
    hints: false,
  },
  output: {
    libraryTarget: 'commonjs2',
    path: path.join(__dirname, '.webpack'),
    filename: '[name].js',
    sourceMapFilename: '[file].map',
  },
};
```
* I then ran `serverless package`

![Screen Shot 2021-03-12 at 11 03 08 AM](https://user-images.githubusercontent.com/14983357/110965788-89441800-8322-11eb-8e5e-89bef1e51bc7.png)

* I then ran `serverless deploy`

![Screen Shot 2021-03-12 at 11 05 54 AM](https://user-images.githubusercontent.com/14983357/110966169-0079ac00-8323-11eb-9e0d-302936e02a6b.png)

![Screen Shot 2021-03-12 at 11 06 00 AM](https://user-images.githubusercontent.com/14983357/110966200-053e6000-8323-11eb-90e6-4f5af0b8eef1.png)


## Todos:


***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO
